File: 프로그래머스/lv2/87946. 피로도/피로도.py
# k = 유저의 현재 피로도
# dungeons = [[던전 별 최소 필요 피로도, 소모 피로도]]
# answer = 최대로 탐험할 수 있는 던전 수

from itertools import permutations

def simulation(k, dungeons):
    cnt = 0
    for dungeon in dungeons:
        if k < dungeon[0]: break
    
        k -= dungeon[1]
        cnt += 1
    return cnt
        
def solution(k, dungeons):
    answer = -1
    
    for case in list(permutations(dungeons, len(dungeons))):
        answer = max(answer, simulation(k, case))
    
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;
/**
k := 피로도
dungeons := [필요 피로도, 소모 필요도]
*/
class Solution {
    static int N, answer;
    public int solution(int k, int[][] dungeons) {
        answer = 0;
        N = dungeons.length;
        boolean[] visited = new boolean[N];
        dfs(0, dungeons, visited, k);
        return answer;
    }
    
    static void dfs(int cnt, int[][] dungeons, boolean[] visited, int HP) {
        if (cnt > answer) answer = cnt;
        
        for(int i = 0; i < N; i++){
            if(!visited[i] && HP >= dungeons[i][0]){
                visited[i] = true;
                dfs(cnt + 1, dungeons, visited, HP - dungeons[i][1]);
                visited[i] = false;
            }
        }
    }
}